## Toronto cluster analysis

In this notebook, you will respond the questions presented for week 3.


Importing libraries


In [1]:
import pandas as pd
from pandas import DataFrame
import numpy as np

Let's create the dataframe

In [2]:
columns_names = ['Postal Code', 'Borough', 'Neighborhood']
df = pd.DataFrame(columns = columns_names)
df

Empty DataFrame
Columns: [Postal Code, Borough, Neighborhood]
Index: []

Let's get raw data fron Wipkipedia and convert it to a dataframe for future manipulation

In [3]:
url = 'https://wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

In [4]:
table = pd.read_html(url)

In [5]:
table[0]

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

Let's get the first table in the web and save the information to a csv file 

In [6]:
table[0].to_csv('Canada_Postal_Code.csv')

Let's visualize the dataframe

In [7]:
table[0]

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

In [8]:
df = table[0]

In [9]:
df.head(10)

Postal Code           Borough                                Neighbourhood
0         M1A      Not assigned                                 Not assigned
1         M2A      Not assigned                                 Not assigned
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
7         M8A      Not assigned                                 Not assigned
8         M9A         Etobicoke      Islington Avenue, Humber Valley Village
9         M1B       Scarborough                               Malvern, Rouge

Cleaning process


First, we should eliminate all Borough with value Not assigned

In [10]:
index_names = df[df['Borough'] == 'Not assigned'].index

In [11]:
df.drop(index_names, inplace=True)

Let's save dataframe to csv

In [12]:
df.to_csv('Canada_Postal_code_clean.csv')

Let's see the new dataframe

In [13]:
df.head(11)

Postal Code           Borough                                Neighbourhood
2          M3A        North York                                    Parkwoods
3          M4A        North York                             Victoria Village
4          M5A  Downtown Toronto                    Regent Park, Harbourfront
5          M6A        North York             Lawrence Manor, Lawrence Heights
6          M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
8          M9A         Etobicoke      Islington Avenue, Humber Valley Village
9          M1B       Scarborough                               Malvern, Rouge
11         M3B        North York                                    Don Mills
12         M4B         East York              Parkview Hill, Woodbine Gardens
13         M5B  Downtown Toronto                     Garden District, Ryerson
14         M6B        North York                                    Glencairn

Let's apply shape to get dataframe dimmensions

In [14]:
df.isna()

Postal Code  Borough  Neighbourhood
2          False    False          False
3          False    False          False
4          False    False          False
5          False    False          False
6          False    False          False
..           ...      ...            ...
160        False    False          False
165        False    False          False
168        False    False          False
169        False    False          False
178        False    False          False

[103 rows x 3 columns]

Therefore, dataframe has three columns and 103 rows


Add coordinates based on Postal Code

In [15]:
df_c = pd.read_csv('Geospatial_Coordinates.csv')

In [16]:
df_c


Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [17]:
import pandas as pd
df.sort_values('Postal Code')

Postal Code      Borough  \
9           M1B  Scarborough   
18          M1C  Scarborough   
27          M1E  Scarborough   
36          M1G  Scarborough   
45          M1H  Scarborough   
..          ...          ...   
98          M9N         York   
107         M9P    Etobicoke   
116         M9R    Etobicoke   
143         M9V    Etobicoke   
152         M9W    Etobicoke   

                                         Neighbourhood  
9                                       Malvern, Rouge  
18              Rouge Hill, Port Union, Highland Creek  
27                   Guildwood, Morningside, West Hill  
36                                              Woburn  
45                                           Cedarbrae  
..                                                 ...  
98                                              Weston  
107                                          Westmount  
116  Kingsview Village, St. Phillips, Martin Grove ...  
143  South Steeles, Silverstone, Humbergate, Jamest...  
152                Northwest, West Humber - Clairville  

[103 rows x 3 columns]

Let's join both dataframes in one 

In [18]:
df_new = df.merge(df_c)
df_new

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

As it was said, we may use Tororonto information only, as far as I understood, I need to include borouggh that contains "Toronto", for example, "Downtown Toronto", "East Toronto" and so on, to do that I will export dataframe to a csv file, AND externally, using Excel I will edit the csv with column Boroughs including Toronto borougHs only...

In [19]:
df_new.to_csv('new_Canada_Postal_code_clean.csv')
df_new.shape

(103, 5)

...new file in my HD in csv format is called Toronto.csv, let's import it into a new datafarme called toronto

In [40]:

toronto = pd.read_csv('Toronto.csv')
toronto

Unnamed: 0 Postal Code           Borough  \
0            2         M5A  Downtown Toronto   
1            4         M7A  Downtown Toronto   
2            9         M5B  Downtown Toronto   
3           15         M5C  Downtown Toronto   
4           19         M4E      East Toronto   
5           20         M5E  Downtown Toronto   
6           24         M5G  Downtown Toronto   
7           25         M6G  Downtown Toronto   
8           30         M5H  Downtown Toronto   
9           31         M6H      West Toronto   
10          36         M5J  Downtown Toronto   
11          37         M6J      West Toronto   
12          41         M4K      East Toronto   
13          42         M5K  Downtown Toronto   
14          43         M6K      West Toronto   
15          47         M4L      East Toronto   
16          48         M5L  Downtown Toronto   
17          54         M4M      East Toronto   
18          61         M4N   Central Toronto   
19          62         M5N   Central Toronto   
20          67         M4P   Central Toronto   
21          68         M5P   Central Toronto   
22          69         M6P      West Toronto   
23          73         M4R   Central Toronto   
24          74         M5R   Central Toronto   
25          75         M6R      West Toronto   
26          79         M4S   Central Toronto   
27          80         M5S  Downtown Toronto   
28          81         M6S      West Toronto   
29          83         M4T   Central Toronto   
30          84         M5T  Downtown Toronto   
31          86         M4V   Central Toronto   
32          87         M5V  Downtown Toronto   
33          91         M4W  Downtown Toronto   
34          92         M5W  Downtown Toronto   
35          96         M4X  Downtown Toronto   
36          97         M5X  Downtown Toronto   
37          99         M4Y  Downtown Toronto   
38         100         M7Y      East Toronto   

                                        Neighbourhood   Latitude  Longitude  
0                           Regent Park, Harbourfront  43.654260 -79.360636  
1         Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
2                            Garden District, Ryerson  43.657162 -79.378937  
3                                      St. James Town  43.651494 -79.375418  
4                                         The Beaches  43.676357 -79.293031  
5                                         Berczy Park  43.644771 -79.373306  
6                                  Central Bay Street  43.657952 -79.387383  
7                                            Christie  43.669542 -79.422564  
8                            Richmond, Adelaide, King  43.650571 -79.384568  
9                        Dufferin, Dovercourt Village  43.669005 -79.442259  
10  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
11                           Little Portugal, Trinity  43.647927 -79.419750  
12                       The Danforth West, Riverdale  43.679557 -79.352188  
13           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
14       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191  
15                     India Bazaar, The Beaches West  43.668999 -79.315572  
16                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
17                                    Studio District  43.659526 -79.340923  
18                                      Lawrence Park  43.728020 -79.388790  
19                                           Roselawn  43.711695 -79.416936  
20                                   Davisville North  43.712751 -79.390197  
21    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307  
22                      High Park, The Junction South  43.661608 -79.464763  
23                  North Toronto West, Lawrence Park  43.715383 -79.405678  
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
25                             Parkdale, Roncesvalles  

.. let's clean new dataframe and delete column Unnamed

In [41]:
toronto = toronto.loc[:, ~toronto.columns.str.contains('^Unnamed')]
toronto

Postal Code           Borough  \
0          M5A  Downtown Toronto   
1          M7A  Downtown Toronto   
2          M5B  Downtown Toronto   
3          M5C  Downtown Toronto   
4          M4E      East Toronto   
5          M5E  Downtown Toronto   
6          M5G  Downtown Toronto   
7          M6G  Downtown Toronto   
8          M5H  Downtown Toronto   
9          M6H      West Toronto   
10         M5J  Downtown Toronto   
11         M6J      West Toronto   
12         M4K      East Toronto   
13         M5K  Downtown Toronto   
14         M6K      West Toronto   
15         M4L      East Toronto   
16         M5L  Downtown Toronto   
17         M4M      East Toronto   
18         M4N   Central Toronto   
19         M5N   Central Toronto   
20         M4P   Central Toronto   
21         M5P   Central Toronto   
22         M6P      West Toronto   
23         M4R   Central Toronto   
24         M5R   Central Toronto   
25         M6R      West Toronto   
26         M4S   Central Toronto   
27         M5S  Downtown Toronto   
28         M6S      West Toronto   
29         M4T   Central Toronto   
30         M5T  Downtown Toronto   
31         M4V   Central Toronto   
32         M5V  Downtown Toronto   
33         M4W  Downtown Toronto   
34         M5W  Downtown Toronto   
35         M4X  Downtown Toronto   
36         M5X  Downtown Toronto   
37         M4Y  Downtown Toronto   
38         M7Y      East Toronto   

                                        Neighbourhood   Latitude  Longitude  
0                           Regent Park, Harbourfront  43.654260 -79.360636  
1         Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
2                            Garden District, Ryerson  43.657162 -79.378937  
3                                      St. James Town  43.651494 -79.375418  
4                                         The Beaches  43.676357 -79.293031  
5                                         Berczy Park  43.644771 -79.373306  
6                                  Central Bay Street  43.657952 -79.387383  
7                                            Christie  43.669542 -79.422564  
8                            Richmond, Adelaide, King  43.650571 -79.384568  
9                        Dufferin, Dovercourt Village  43.669005 -79.442259  
10  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
11                           Little Portugal, Trinity  43.647927 -79.419750  
12                       The Danforth West, Riverdale  43.679557 -79.352188  
13           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
14       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191  
15                     India Bazaar, The Beaches West  43.668999 -79.315572  
16                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
17                                    Studio District  43.659526 -79.340923  
18                                      Lawrence Park  43.728020 -79.388790  
19                                           Roselawn  43.711695 -79.416936  
20                                   Davisville North  43.712751 -79.390197  
21    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307  
22                      High Park, The Junction South  43.661608 -79.464763  
23                  North Toronto West, Lawrence Park  43.715383 -79.405678  
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
25                             Parkdale, Roncesvalles  43.648960 -79.456325  
26                                         Davisville  43.704324 -79.388790  
27                     University of Toronto, Harbord  43.662696 -79.400049  
28                                 Runnymede, Swansea  43.651571 -79.484450  
29                        Moore Park, Summerhill East  43.689574 -79.383160  
30          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
31  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -

Therefore, we have a new dataframe with Postal Code, Borough, Neighborwooh and latitude and longitudes 
Now, Let's go into visualization process, first at all, let's install folium and webbrowser libraries...

In [26]:
!conda install -c conda-forge folium=0.5.0 --yes 
import folium 
import webbrowser

Solving environment: ...working... done

# All requested packages already installed.



let's get Toronto coordinates by Google and generate our object map and named it as traffic_map

In [29]:
latitude = 43.651070
longitude = -79.347015

traffic_map = folium.Map(location=[latitude, longitude], zoom_start=4)

now, let's create the map instance m using the location...

In [30]:
m = folium.Map(location=[latitude, longitude])

and let's add the makers to map using method CircleMaker ...


In [39]:
for lat, lng, label in zip(toronto['Latitude'], toronto['Longitude'], toronto['Borough']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(m)
m

Finally, we have visualized the Toronto boroughs in the map.